### PR TITLE
SMT Backend: move undefined memory behavior modelling to firrtl IR level

### PR DESCRIPTION
**Pull request by ekiwi**
**tl;dr** With this PR the smt backend now supports memories with more than two write ports and the conservative memory modelling can be selectively turned off with a new annotation.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you update the FIRRTL spec to include every new feature/behavior?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement
- code refactoring

#### API Impact

- this adds the `UndefinedMemoryBehaviorPass` and `UndefinedMemoryBehaviorOptions` annotation as part of the experimental SMT backend

#### Backend Code Generation Impact

- small changes in the emitted bor2/SMT, however the functionality should be the same

#### Desired Merge Strategy

- squash

#### Release Notes
 n/a

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/firrtl/transforms/DeadCodeElimination.scala [hunk 1]
```diff
@@ -126,6 +127,11 @@ class DeadCodeElimination extends Transform with RegisteredTransform with Depend
         val node = LogicNode(mod.name, name)
         depGraph.addVertex(node)
         Seq(clock, reset, init).flatMap(getDeps(_)).foreach(ref => depGraph.addPairWithEdge(node, ref))
+      case DefRandom(_, name, _, clock, en) =>
+        val node = LogicNode(mod.name, name)
+        depGraph.addVertex(node)
+        val inputs = List() ++ clock ++ en
```
Comment (jackkoenig) [inline src/main/scala/firrtl/transforms/DeadCodeElimination.scala:133]:
```suggestion
        val inputs = clock ++: en ++: Nil
```

File: src/main/scala/firrtl/Utils.scala [hunk 2]
```diff
@@ -868,6 +868,58 @@ object Utils extends LazyLogging {
   def maskBigInt(value: BigInt, width: Int): BigInt = {
     value & ((BigInt(1) << width) - 1)
   }
+
+  def isLiteral(e: Expression): Boolean = e match {
```
[the remark this answers is not in the document]
Comment (jackkoenig) [reply src/main/scala/firrtl/Utils.scala:872]: I think noting that it does more than `.isInstanceOf[Literal]` might be helpful. I'd say `isLiteral` is a little more self explanatory but explaining that say `and` does some constant propagation is useful to note.

[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/Utils.scala:872]: So like `/** Applies the firrtl And primop. Automatically constant propagates when one of the expressions is True or False. */`?

File: src/main/scala/firrtl/backends/experimental/smt/FirrtlToTransitionSystem.scala [hunk 3]
```diff
@@ -247,233 +250,116 @@ private class ModuleToTransitionSystem extends LazyLogging {
...
+  }
+
+  private[firrtl] val randoms = mutable.LinkedHashMap[String, BVSymbol]()
+  private def makeRandom(baseName: String, width: Int): BVExpr = {
+    // TODO: actually ensure that there cannot be any name clashes with other identifiers
+    val suffixes = Iterator(baseName) ++ (0 until 200).map(ii => baseName + "_" + ii)
+    val name = suffixes.map(s => "RANDOM." + s).find(!randoms.contains(_)).get
+    val sym = BVSymbol(name, width)
+    randoms(name) = sym
+    sym
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/backends/experimental/smt/FirrtlToTransitionSystem.scala:361]: This is old code that is going to be removed by a followup PR. Since we are moving all randomness modelling out of the SMT backend and into passes that add `DefRandom` nodes, the backend will no longer need to generate names.

File: src/main/scala/firrtl/WIR.scala [hunk 4]
```diff
@@ -13,6 +13,7 @@ trait Kind
 case object WireKind extends Kind
 case object PoisonKind extends Kind
 case object RegKind extends Kind
+case object RandomKind extends Kind
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/WIR.scala:16]: I do not like the term Poison since afaik it isn't the same as poison in LLVM.
I am open to changing `Random` to `Any` or `Arbitrary` or `AnySeq`.

File: src/main/scala/firrtl/backends/experimental/smt/random/UndefinedMemoryBehaviorPass.scala [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+  override def invalidates(a: Transform) = a match {
+    // this pass might destroy SSA form, as we add a wire for the data field of every read port
+    case _: RemoveWires => true
+    // TODO: should we add some optimization passes here? we could be generating some dead code.
+    case _ => false
+  }
+
+  override protected def execute(state: CircuitState): CircuitState = {
+    val opts = state.annotations.collect { case o: UndefinedMemoryBehaviorOptions => o }
+    assert(opts.size < 2, s"Multiple options: $opts")
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/backends/experimental/smt/random/UndefinedMemoryBehaviorPass.scala:49]: I can change that.

File: src/main/scala/firrtl/Utils.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/Utils.scala:873]: done

File: src/main/scala/firrtl/backends/experimental/smt/random/DefRandom.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package firrtl.backends.experimental.smt.random
+
+import firrtl.ir._
+
+/** Named source of random values. If there is no clock expression, than it will be clocked by the global clock. */
+case class DefRandom(
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/backends/experimental/smt/random/DefRandom.scala:9]: Alternative names that I would be fine with: `DefAny`, `DefArbitrary`, `DefAnySeq`

I don't want to use the term `Poison` since that is LLVM terminology for something similar but not quite the same as what we are talking about here. So I believe there is too much room for confusion.

File: src/main/scala/firrtl/backends/experimental/smt/random/UndefinedMemoryBehaviorPass.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/firrtl/backends/experimental/smt/random/UndefinedMemoryBehaviorPass.scala]: done

**Review by jackkoenig — APPROVED** [review]
LGTM, we can bikeshed more about the name before 1.5 is released
